### ISSUE TITLE
Add Quasar plugin object into TypeScript typings

### ISSUE THREAD

**Issue by IlCallo** [quasarframework/quasar#4584]
**Describe the bug**
Quasar typings don't cover the plugin object, which is required to mount Quasar during testing (in Jest, for example).

The `Quasar` plugin object is defined here:
https://github.com/quasarframework/quasar/blob/58c0ac7aea433cef13bccac7b61f2571bad9f8d0/ui/src/index.common.js#L21
and thus included into the generated `quasar.common.js`. This allow it to work when using JS and including it like `import { Quasar } from 'quasar'`, but as soon as you enable TypeScript you'll get an error because it has not been defined into typings.

`Quasar` should be of type `PluginObject<T>`, where `T` are the optional parameters of the plugin.

**Expected behavior**
Based on the plugin [installation function](https://github.com/quasarframework/quasar/blob/dev/ui/src/install.js), the typings to add should probably be:

```typescript
import { PluginObject } from 'vue';

export interface QuasarPluginOptions {
  lang: ???,
  config: ???,
  iconSet: ???,
  components: ???,
  directives: ???,
  plugins: ???
}

export const Quasar = PluginObject<QuasarPluginOptions>;
```
I'm not sure about the types of the properties, that's why I just put `???` everywhere.

**Platform (please complete the following information):**
OS: W10
Node: 10.x
Yarn: 1.16

**Additional informations**
Shim to workaround the problem.

<details>
<summary>Easy way</summary>
Create a new `quasar.d.ts` in `src/` and paste this inside

```typescript
import { PluginObject } from 'vue';

declare module 'quasar' {
  interface QuasarPluginOptions {
    lang: any;
    config: any;
    iconSet: any;
    components: any;
    directives: any;
    plugins: any;
  }

  const Quasar: PluginObject<QuasarPluginOptions>;
}
```

</details>

---

<details>
<summary>Complete way</summary>
Create a new `quasar-shims` folder in `src/` and create these three files

<details>
<summary>utils.ts</summary>

```ts
export type StringDictionary<T extends string> = Required<{ [index in T]: string }>;
```

</details>

<details>
<summary>lang.ts</summary>

```ts
import { StringDictionary } from './utils';

type QuasarLanguageGeneralLabel =
  | 'clear'
  | 'ok'
  | 'cancel'
  | 'close'
  | 'set'
  | 'select'
  | 'reset'
  | 'remove'
  | 'update'
  | 'create'
  | 'search'
  | 'filter'
  | 'refresh';
type QuasarLanguageTableLabel =
  | 'noData'
  | 'noResults'
  | 'loading'
  | 'recordsPerPage'
  | 'allRows'
  | 'columns';
type QuasarLanguageEditorLabel =
  | 'url'
  | 'bold'
  | 'italic'
  | 'strikethrough'
  | 'underline'
  | 'unorderedList'
  | 'orderedList'
  | 'subscript'
  | 'superscript'
  | 'hyperlink'
  | 'toggleFullscreen'
  | 'quote'
  | 'left'
  | 'center'
  | 'right'
  | 'justify'
  | 'print'
  | 'outdent'
  | 'indent'
  | 'removeFormat'
  | 'formatting'
  | 'fontSize'
  | 'align'
  | 'hr'
  | 'undo'
  | 'redo'
  | 'header1'
  | 'header2'
  | 'header3'
  | 'header4'
  | 'header5'
  | 'header6'
  | 'paragraph'
  | 'code'
  | 'size1'
  | 'size2'
  | 'size3'
  | 'size4'
  | 'size5'
  | 'size6'
  | 'size7'
  | 'defaultFont';

type QuasarLanguageTreeLabel = 'noNodes' | 'noResults';
type QuasarLanguageDayTuple = [
  string,
  string,
  string,
  string,
  string,
  string,
  string
];
type QuasarLanguageMonthTuple = [
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string
];

export interface QuasarLanguage {
  isoName: string;
  nativeName: string;
  rtl?: boolean;
  label: StringDictionary<QuasarLanguageGeneralLabel>;
  date: {
    days: string[]; // Should be QuasarLanguageDayTuple, but string splitting doesn't allow to check array length
    daysShort: string[]; // Should be QuasarLanguageDayTuple, but string splitting doesn't allow to check array length
    months: string[]; // Should be QuasarLanguageMonthTuple, but string splitting doesn't allow to check array length
    monthsShort: string[]; // Should be QuasarLanguageMonthTuple, but string splitting doesn't allow to check array length
    firstDayOfWeek: number;
    format24h: boolean;
    // Could not understand what exactly 'model' is, modeled after its usage
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    headerTitle?: (
      date: Date,
      model: { year: number; month: number; day: number },
    ) => string;
  };
  table: StringDictionary<QuasarLanguageTableLabel> & {
    selectedRecords: (rows: number) => string;
    pagination: (start: number, end: number, total: number) => string;
  };
  editor: StringDictionary<QuasarLanguageEditorLabel>;
  tree: StringDictionary<QuasarLanguageTreeLabel>;
}
```

</details>

<details>
<summary>icon-set.ts</summary>

```ts
import { StringDictionary } from './utils';

type QuasarIconSetType = 'positive' | 'negative' | 'info' | 'warning';
type QuasarIconSetArrow = 'up' | 'right' | 'down' | 'left' | 'dropdown';
type QuasarIconSetChevron = 'left' | 'right';
type QuasarIconSetColorPicker = 'spectrum' | 'tune' | 'palette';
type QuasarIconSetPullToRefresh = 'icon';
type QuasarIconSetCarousel = 'left' | 'right' | 'navigationIcon' | 'thumbnails';
type QuasarIconSetChip = 'remove' | 'selected';
type QuasarIconSetDatetime = 'arrowLeft' | 'arrowRight' | 'now' | 'today';
type QuasarIconSetEditor =
  | 'bold'
  | 'italic'
  | 'strikethrough'
  | 'underline'
  | 'unorderedList'
  | 'orderedList'
  | 'subscript'
  | 'superscript'
  | 'hyperlink'
  | 'toggleFullscreen'
  | 'quote'
  | 'left'
  | 'center'
  | 'right'
  | 'justify'
  | 'print'
  | 'outdent'
  | 'indent'
  | 'removeFormat'
  | 'formatting'
  | 'fontSize'
  | 'align'
  | 'hr'
  | 'undo'
  | 'redo'
  | 'header'
  | 'code'
  | 'size'
  | 'font';
type QuasarIconSetExpansionItem = 'icon' | 'denseIcon';
type QuasarIconSetFab = 'icon' | 'activeIcon';
type QuasarIconSetField = 'clear' | 'error';
type QuasarIconSetPagination = 'first' | 'prev' | 'next' | 'last';
type QuasarIconSetRating = 'icon';
type QuasarIconSetStepper = 'done' | 'active' | 'error';
type QuasarIconSetTabs = 'left' | 'right' | 'up' | 'down';
type QuasarIconSetTable = 'arrowUp' | 'warning' | 'prevPage' | 'nextPage';
type QuasarIconSetTree = 'icon';
type QuasarIconSetUploader =
  | 'done'
  | 'clear'
  | 'add'
  | 'upload'
  | 'removeQueue'
  | 'removeUploaded';

export interface QuasarIconSet {
  name: string;
  type: StringDictionary<QuasarIconSetType>;
  arrow: StringDictionary<QuasarIconSetArrow>;
  chevron: StringDictionary<QuasarIconSetChevron>;
  colorPicker: StringDictionary<QuasarIconSetColorPicker>;
  pullToRefresh: StringDictionary<QuasarIconSetPullToRefresh>;
  carousel: StringDictionary<QuasarIconSetCarousel>;
  chip: StringDictionary<QuasarIconSetChip>;
  datetime: StringDictionary<QuasarIconSetDatetime>;
  editor: StringDictionary<QuasarIconSetEditor>;
  expansionItem: StringDictionary<QuasarIconSetExpansionItem>;
  fab: StringDictionary<QuasarIconSetFab>;
  field: StringDictionary<QuasarIconSetField>;
  pagination: StringDictionary<QuasarIconSetPagination>;
  rating: StringDictionary<QuasarIconSetRating>;
  stepper: StringDictionary<QuasarIconSetStepper>;
  tabs: StringDictionary<QuasarIconSetTabs>;
  table: StringDictionary<QuasarIconSetTable>;
  tree: StringDictionary<QuasarIconSetTree>;
  uploader: StringDictionary<QuasarIconSetUploader>;
}
```

</details>

Then create a new `quasar.d.ts` in `src/` and paste this inside

```ts
// Shim for Quasar PluginObject
// See https://github.com/quasarframework/quasar/issues/4584
import { PluginObject } from 'vue';

import { QuasarIconSet } from './quasar-shims/icon-set';
import { QuasarLanguage } from './quasar-shims/lang';

declare module 'quasar' {
  interface QuasarPluginOptions {
    lang: QuasarLanguage;
    config: any;
    iconSet: QuasarIconSet;
    components: any;
    directives: any;
    plugins: any;
  }

  const Quasar: PluginObject<QuasarPluginOptions>;
}
```

</details>

**Comment by outofmemoryagain** [comment]
@IlCallo I can pretty easily get the shim to be part of the generated types.  I can also add the components, directive, and plugins with an actual list that represent those values in a type safe way.  I'm not 100% sure at the moment about lang, config, and iconSet.  They appear to be vue plugins themselves which I'm not sure about represting with a Type in this case.  I will start a PR, and reference here.

**Comment by outofmemoryagain** [comment]
@IlCallo take a look at the following [branch](https://github.com/outofmemoryagain/quasar/tree/fix_4584).  It should generate a type that is usable for the test extension.  lang, config, and iconSet are of type any.  I'm not entirely sure about those yet, but I can look into those if they should be better typed as well.  If you need help generating the type def file let me, but it looks like you have it figured out.  Let me know if the generated type fills your need.

**Comment by IlCallo** [comment]
Pretty hard to say if it's ok, that file isn't much clear to read at first sight. I guess it _looks_ right 🤔
Is it possible to check the generated file?

- for `lang`, if there isn't a way to reduce its type to a limited set of strings (and I guess it's not that easy or desirable), it should probably be of type `string`
- for `iconSet`, I guess it should be an union type with all icon sets names accepted by Quasar, which the documentation point out to be the names of [files in this folder](https://github.com/quasarframework/quasar/tree/dev/ui/icon-set) (more precisely, I think it's the `name` property of default objects exported by these files, and file names are defined the same as consequence). Giving that you are building it up with a node script, you can directly read all files in the folder and compose the type dynamically, to make it future-proof
- for `config`.... No idea 😓 but it would be a good thing to have it fully typed

PS: don't you have a function to write multiple lines starting from a multiline template string in that file? That seems an extremely verbose why of composing a text-based file 🤔

**Comment by outofmemoryagain** [comment]
@IlCallo Here is the generated file. Line 1825-1972 and Line 2113 are of where the changes are for this type.
[index.d.ts.txt](https://github.com/quasarframework/quasar/files/3382472/index.d.ts.txt)

As far as lang, iconSt and config go, when looking at the code it looks like those properties on the plugin are plugins themselves and are used to initialize the lang, iconSet and config used by Quasar internally.  But I would need to look at it with fresher eyes than I had last night.

**Comment by outofmemoryagain** [comment]
@IlCallo, so I took a look at this with fresher eyes and understand what this represents a lot better.  It helps to read the issue in a bit more detail... sometimes just digging into the code helps it click a bit better.
* The language [pack documentation](https://quasar.dev/options/quasar-language-packs#Vue-CLI) states shows the following for setting the lang option:
```
import langDe from 'quasar/lang/de'
import { Quasar } from 'quasar'
...
Vue.use(Quasar, {
  ...
  lang: langDe
})
```
* The [iconSet document](https://quasar.dev/options/quasar-icon-sets#Vue-CLI-Way) shows the following for setting the iconSet Option
```
import iconSet from 'quasar/icon-set/fontawesome-v5'
import { Quasar } from 'quasar'
...
Vue.use(Quasar, {
  iconSet: iconSet
})
```
And after reviewing the code in 'quasar/lang/de' and quasar/icon-set/fontawesome-v5 they are both contain a json structure that would be difficult to add a type for since it the structure looks fairly dynamic.  I could create types that represent both, but I don't think there is much value here in this case beyond defining them as any type, which the current PR does.  I could consider as you mentioned read each one of these files and derive and generated a type and then declare a const or something along those lines for each file.  That may be outside the scope of the time I have available.

**Comment by IlCallo** [comment]
I gave a try and created the needed interfaces.
I'm gonna update the initial post with new shims in a while, then it's up to you if you want to use them or not, I'll just leave them there if someone wants to add them in their projects.

What I found while testing the interface on languages:
- https://github.com/quasarframework/quasar/blob/9e40b6a066c38e55670fd808a57c72cf5a1e1ae1/ui/lang/eo.js#L88 should be `noNodes`
- https://github.com/quasarframework/quasar/blob/9e40b6a066c38e55670fd808a57c72cf5a1e1ae1/ui/lang/it.js#L88 should be `noNodes`
- https://github.com/quasarframework/quasar/blob/9e40b6a066c38e55670fd808a57c72cf5a1e1ae1/ui/lang/fi.js#L27-L31 exists only here and is never used elsewhere, so I guess it should be removed
- https://github.com/quasarframework/quasar/blob/9e40b6a066c38e55670fd808a57c72cf5a1e1ae1/ui/lang/fr.js#L88 should be `noNodes`
- https://github.com/quasarframework/quasar/blob/9e40b6a066c38e55670fd808a57c72cf5a1e1ae1/ui/lang/nb-no.js#L31 exists only here and seems like it's never used elsewhere (not with a direct reference, at least)

No strange things in icon-sets.

**Comment by IlCallo** [comment]
On regards to the generated file: I think that in the way you put `components`, `directives` and `plugins` you are requiring to add _**all of them**_, while I though you can just add some of them (the ones you need for testing).
You probably want to use a mapped object or set all of the options as optional like this

```ts
interface QuasarComponents {
  QAjaxBar?: VueConstructor<QAjaxBar>;
  QAvatar?: VueConstructor<QAvatar>;
  /* ... */
}

// ...
export interface QuasarPluginOptions {
   // ...
   components: QuasarComponents,
   // ...
}
```

I think there's a way to make it a mapped object, but it may not be safe if you apply aggressive minification which mess up with constants and class names.

**Comment by outofmemoryagain** [comment]
Yeah, looked into `config` a little, but had difficulty tracking any documented example of specifying that property.  I believe it is used by the CLI, but I haven't been able to determine exactly what is passes there.  I even added a console log to output the options in the plugin to examine what is in the config object, and always {}.  Although I haven'r tried one of the other modes (pwa, cordova, electron..) while logging the object.

Also, I can easily make each component optional.  Thanks for catching that one.

As far as the suggested shim, it would not work if there already a type with those properties specified on them.  You can not used Type Declaration merging to change the type of an existing property, you can only add new one.

**Comment by outofmemoryagain** [comment]
I can add the QLanguage and QIconSet types as statically defined types into /ui/types, but I am slightly concerned about maintainability of them.  I can discuss it over with the core team and get their thoughts on it.

**Comment by IlCallo** [comment]
> As far as the suggested shim, it would not work if there already a type with those properties specified on them. You cannot use Type Declaration merging to change the type of an existing property, you can only add new ones (thus you are merging to separate declaration files in to a single declaration).

I'm not understanding what you are referencing here 🤔 

> I can add the QLanguage and QIconSet types as statically defined types into /ui/types, but I am slightly concerned about maintainability of them. I can discuss it over with the core team and get their thoughts on it.

Yeah, they are hard to keep up to date if you are not actually using them into the core...

**Comment by outofmemoryagain** [comment]
I just had a chat with @nothingismagick and
>well the thing is that those interfaces are really pretty obscured from direct interaction
>i mean its like a configuration layer thing
>so with lang that can change at runtime
>but icons are (for the moment) still set at build time

I think we are safe statically defining a QIconSet type but not so much a QLang type.

>>As far as the suggested shim, it would not work if there already a type with those properties specified >>on them. You cannot use Type Declaration merging to change the type of an existing property, you >>can only add new ones (thus you are merging to separate declaration files in to a single declaration).

> I'm not understanding what you are referencing here 🤔

This is referring your shim suggestion for setting a more typed version of the QuasarPluginOptions type than is defined by the framework.
So if the framework defines
```
declare module 'quasar' {
  interface QuasarPluginOptions {
    lang: any;
    config: any;
    iconSet: QIconSet;
    components: QComponents;
    directives: QDirectives;
    plugins: QPlugins;
  }

  const Quasar: PluginObject<QuasarPluginOptions>;
}
```
You would not able to change the `lang: any` to `lang: QLanguage` in a shim.

**Comment by outofmemoryagain** [comment]
I will make a new PR tonight to make the components optional :-) and add a type for lang

**Comment by rstoenescu** [comment]
Fix will be available in 1.0.5.

**Comment by outofmemoryagain** [comment]
@rstoenescu when is that planned?  There is one slight issue @IlCallo just pointed out with the PR that I would like to get fixed before it is released.

**Comment by rstoenescu** [comment]
@outofmemoryagain I'll do a release with 1.0.5 tomorrow. Will wait for your PR. Thanks for taking care of this!

**Comment by outofmemoryagain** [comment]
ok, I will try to get it done tonight

**Comment by IlCallo** [comment]
> I will make a new PR tonight to make the components optional

You probably also want to make `Quasar` of type `PluginObject<Partial<QuasarPluginOptions>>`, now that I think about it: the installation function defaults to an empty object, so it manages internally its defaults I guess. I updated the shim in the first post.

> You would not able to change the lang: any to lang: QLanguage in a shim.

I can live without it 😁 
Also, it's pretty easy to generate a custom type with my QuasarLanguage shim type and type-cast `Vue.use()`

```ts
type BetterQuasarPluginOptions = Omit<Partial<QuasarPluginOptions>, 'lang'> & {lang?: QuasarLanguage};
Vue.use(Quasar, { /* 'lang' have type 'QuasarLanguage | undefined', which I'll probably never use anyway 🤷‍♂  */ })
```

**Comment by rstoenescu** [comment]
Thank you both for collaborating and finding a solution!
Fix will be available in 1.0.5.
